### ISSUE TITLE
Enum value not assingable to typeof Enum

### ISSUE THREAD

**Issue by grissius** [Microsoft/TypeScript#30080]
**TypeScript Version:**  3.4.0-dev.20190223

**Search Terms:** enum, typeof enum, not assignable, dynamic enums

**Code**

```ts
enum E {
  Foo = 'Foo'
}

const castedE = E as typeof E;
const fun = (e: typeof E) => e;
fun(castedE.Foo) //  Argument of type 'E' is not assignable to parameter of type 'typeof E'.
```
Agree it is not a usual usecase, but from the error message it seems like a bug. It is encountered when user tries to dynamically temper with enums.

**Expected behavior:** No error

**Actual behavior:**
```
 Argument of type 'E' is not assignable to parameter of type 'typeof E'.
```

**Playground Link:** [link](https://www.typescriptlang.org/play/index.html#src=enum%20E%20%7B%0D%0A%20%20Foo%20%3D%20'Foo'%0D%0A%7D%0D%0A%0D%0Aconst%20castedE%20%3D%20E%20as%20typeof%20E%3B%0D%0Aconst%20fun%20%3D%20(e%3A%20typeof%20E)%20%3D%3E%20e%3B%0D%0Afun(castedE.Foo))

**Related Issues:**
https://github.com/Microsoft/TypeScript/issues/28102
https://github.com/Microsoft/TypeScript/issues/17690
https://github.com/Microsoft/TypeScript/issues/22781
(not sure how much related)

**Comment by dragomirtitian** [comment]
`typeof E` represent the type of the container object for the enum. At runtime, enums have an object that contains the enum values (unless you have a `const` enum). So the type of `typeof E` would be equivalent to `Record<"Foo", E>`.

Depending on what you want to achieve you can fix the error either by passing in the enum:

```ts
const fun = (e: typeof E) => e;
fun(castedE)
```

Or by chaining the type of the parameter: 

```ts
const fun = (e: E) => e;
fun(castedE.Foo)
```

or relative to `castedE`

```ts

const castedE = E; // no cast really necessary
const fun = (e: typeof castedE[keyof typeof castedE]) => e;
fun(castedE.Foo)
```

**Comment by grissius** [comment]
I want to pass the value, not the container, so only either of the two last options are relevant. I wanted to avoid exposing the original enum `E`, only the modified version, which means I have to take the last one. Thanks for explanation, it solved my issue.

However, I still think that the fact that `castedE` in  `const castedE = E as typeof E;` behaves differently and I daresay in a non intuitive way might be an issue for other users.
